contracts/governance/Vesting/OriginInvestorsClaim.sol

```diff
@@ -0,0 +1,178 @@
+		emit AdminAdded(_admin);
+	}
+
+	function removeAdmin(address _admin) public onlyOwner {
+		admins[_admin] = false;
+		emit AdminRemoved(_admin);
+	}
+
+	/**
+	 * @notice in case we have unclaimed tokens or in emergency case
+	 */
+	function authorizedBalanceWithdraw(address toAddress) public onlyAuthorized {
+		require(
+			SOVToken.transfer(toAddress, SOVToken.balanceOf(address(this))),
+			"OriginInvestorsClaim::authorizedTransferBalance: transfer failed"
+		);
+	}
+
+	/**
+	 * @notice should ne called after the investors list setup completed
+	 */
+	function setInvestorsAmountsListInitialized() public onlyAuthorized notInitialized {
+		require(
+			SOVToken.balanceOf(address(this)) >= totalAmount,
+			"OriginInvestorsClaim::setInvestorsAmountsList: the contract is not enough financed"
+		);
+
+		investorsListInitialized = true;
+
+		emit InvestorsAmountsListInitialized(investorsQty, totalAmount);
+	}
+
+	/**
+	 *  @notice the contract should be approved or transferred necessary amount of SOV prior to calling the function
+	 *  @param investors is the list of investors addresses to add to the list. Duplicates will be skipped.
+	 *  @param claimAmounts is the list of amounts for investors investors[i] will receive claimAmounts[i] of SOV
+	 */
+	function appendInvestorsAmountsList(address[] calldata investors, uint256[] calldata claimAmounts)
+		external
+		onlyAuthorized
+		notInitialized
+	{
+		uint256 subQty;
+		uint256 sumAmount;
+		require(
+			investors.length == claimAmounts.length,
+			"OriginInvestorsClaim::appendInvestorsAmountsList: investors.length != claimAmounts.length"
+		);
+
+		for (uint256 i = 0; i < investors.length; i++) {
+			if (investorsAmountsList[investors[i]] == 0) {
+				investorsAmountsList[investors[i]] = claimAmounts[i];
+				sumAmount = sumAmount.add(claimAmounts[i]);
+			} else {
+				subQty = subQty.add(1);
+			}
+		}
+
+		investorsQty = investorsQty.add(investors.length.sub(subQty));
+		totalAmount = totalAmount.add(sumAmount);
+		emit InvestorsAmountsListAppended(investors.length.sub(subQty), sumAmount);
+	}
+
+	function claim() external onlyWhitelisted initialized {
+		if (now < vestingTerm) {
+			createVesting();
+		} else {
+			transfer();
+		}
+	}
+
+	function createVesting() internal {
+		uint256 cliff = vestingTerm.sub(now);
+		uint256 duration = cliff;
+		uint256 amount = investorsAmountsList[msg.sender];
+		address vestingContractAddress;
+
+		vestingContractAddress = vestingRegistry.getVesting(msg.sender);
+		require(vestingContractAddress == address(0), "OriginInvestorsClaim::withdraw: the claimer has an active vesting contract");
+
+		delete investorsAmountsList[msg.sender];
+
+		vestingRegistry.createVesting(msg.sender, amount, cliff, duration);
+		vestingContractAddress = vestingRegistry.getVesting(msg.sender);
+		require(SOVToken.transfer(address(vestingRegistry), amount), "OriginInvestorsClaim::withdraw: SOV transfer failed");
+		vestingRegistry.stakeTokens(vestingContractAddress, amount);
```

Comment:
You're right, initially VestingRegistry.stakeTokens was a part of VestingRegistry.createVesting, I had to split them because of block gas limit.
We can use direct invocation here:
`IVesting(vestingContractAddress).stakeTokens(amount);`
and approve instead of transfer one line before:
`SOVToken.approve(vestingContractAddress, amount);`

I'm not sure it's a right time to do this change now unless we can save significant amount of gas.